services/src/LinkLossService/LinkLossService.cpp

```diff
@@ -0,0 +1,110 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mbed-os-experimental-ble-services/services/inc/LinkLossService.h"
+
+LinkLossService::LinkLossService(BLE &ble, events::EventQueue &event_queue, ChainableGapEventHandler &chainable_gap_event_handler) :
+    _ble(ble),
+    _event_queue(event_queue),
+    _chainable_gap_event_handler(chainable_gap_event_handler),
+    _alert_level_char(GattCharacteristic::UUID_ALERT_LEVEL_CHAR, &_alert_level)
+{
+}
+
+LinkLossService::~LinkLossService()
+{
+}
+
+ble_error_t LinkLossService::init()
+{
+    GattCharacteristic *charTable[] = { &_alert_level_char };
+    GattService         linkLossService(GattService::UUID_LINK_LOSS_SERVICE, charTable, 1);
+
+    _alert_level_char.setWriteAuthorizationCallback(this, &LinkLossService::onDataWritten);
+
+    ble_error_t bleError = _ble.gattServer().addService(linkLossService);
+
+    if (!bleError) {
+        _chainable_gap_event_handler.addEventHandler(this);
+    }
+
+    return bleError;
+}
+
+void LinkLossService::set_event_handler(EventHandler* handler)
+{
+    alert_handler = handler;
+}
+
+void LinkLossService::set_level(AlertLevel level)
+{
+    _alert_level = level;
+}
+
+void LinkLossService::set_timeout(std::chrono::milliseconds timeout)
+{
+    _alert_timeout = timeout;
+}
+
+LinkLossService::AlertLevel LinkLossService::get_level()
+{
+    return _alert_level;
+}
+
+void LinkLossService::stop_alert()
+{
+    _event_queue.cancel(_event_queue_handle);
+    if (in_alert) {
+        in_alert = false;
+        if (alert_handler) {
+            alert_handler->on_alert_end();
+        }
+    }
+}
+
+void LinkLossService::onConnectionComplete(const ble::ConnectionCompleteEvent &event)
+{
+    stop_alert();
+}
+
+void LinkLossService::onDisconnectionComplete(const ble::DisconnectionCompleteEvent &event)
+{
+    AlertLevel level = get_level();
+    if (event.getReason() == ble::disconnection_reason_t::CONNECTION_TIMEOUT && level != AlertLevel::NO_ALERT) {
+        in_alert = true;
+        alert_handler->on_alert_requested(level);
+        if (_alert_timeout > std::chrono::milliseconds(0)) {
+            _event_queue_handle = _event_queue.call_in(_alert_timeout, [this] { stop_alert(); });
+        }
+    }
+}
+
+void LinkLossService::onDataWritten(GattWriteAuthCallbackParams *write_request)
+{
+    const uint8_t level = *write_request->data;
+
+    if (write_request->offset != 0) {
+        write_request->authorizationReply = AUTH_CALLBACK_REPLY_ATTERR_INVALID_OFFSET;
+        return;
+    }
+
```

Comment:
Please check the size of the write request too.

Reply:
is all this checking really necessary? doesn't gatt server already check this?

Reply:
also see my comment below

Reply:
You are right, the characteristic has a fixed length and a size of 1. If the offset and length are invalid then it should be rejected by the server before the invocation of this callback.